tests/integration/psync2-master-restart.tcl

```diff
@@ -30,15 +30,17 @@ start_server {} {
     createComplexDataset $master 1000
 
     test "PSYNC2: Partial resync after Master restart using RDB aux fields" {
+        set offset [status $master master_repl_offset]
         wait_for_condition 500 100 {
-            [status $master master_repl_offset] == [status $replica master_repl_offset] &&
-            [status $master master_repl_offset] == [status $sub_replica master_repl_offset]
+            [string match "*slave0:*,offset=$offset,*" [$master info replication]] &&
+            $offset == [status $replica master_repl_offset] &&
+            $offset == [status $sub_replica master_repl_offset]
```

Comment:
why did you need to change this?
it already changed `repl-ping-replica-period` to high value?

please explain what was the problem.
i'm not sure if comparing to a static pre-sampled offset is right, and if we do that maybe we need to use `>=` (which i think can hide bugs)

Reply:
I changed this because REPLCONF GETACK affects the offset and this broke the asserts later in the test case. It seems that the offset is incremented by exactly 37 when REPLCONF GETACK is sent.

If the purpose of the test case has anything to do with the master not knowing that the replica is up to date (because it is; it's checked in this `wait_for_condition` loop), then we can't do this, but OTOH with this PR the master will send GETACK during shutdown and wait for the replicas anyway, so the high value of `repl-ping-replica-period` isn't effective anyway.

My though was that `>=` would hide bugs, so I didn't want to do that.

What if we use `shutdown now` to restart the master? or equivalently `debug restart`. That prevents the REPLCONF GETACK...

Reply:
I did the latter and used `$master debug restart` instead of `restart_server 0 true false`. It uses SHUTDOWN_NOW, preserving the old behaviour of the test case.

Reply:
these tests used to use DEBUG RESTART in the past.
it upsets valgrind and causes other test suite issues, which is why i introduced restart_server and used it everywhere.
let's add an argument to restart_server to specify how it should be shut down.

Reply:
after thinking a bit more about it, it occurred to me that maybe this exposes a bigger problem.
maybe that extra GETACK feature can cause real issues of a repl offset mismatch between the RDB file and the replicas.
specifically if the timeout is short.

but looking again ad your change (which you now reverted), i understand that at least in that test, this extra GETACK feature didn't cause any replication issues (PSYNC still succeeded). all it did is fail an assertion in the test, right?

so considering that, maybe i'm ok with that change (which you reverted).
but also, maybe we do want to run this test on both modes (SIGTERM, and SHUTDOWN NOW).
maybe we should modify one test to use one approach, and the other to use the other approach (just in order to avoid running 4 tests instead of 2)

Reply:
That's right, all it did was fail an assertion in the tests. The assertion checked that the offset reported by the replica matches the offset reported by the master, but it didn't check that the master had received the offset (ACK) from the replica. If it hadn't received the ACK, it would wait for replicas and send GETACK to speed up the ACK.

So, I agree, I don't think the GETACK can cause any serious issues. If the replicas are already up to date, it will not be sent. If they're not, the GETACK will speed up the shutdown.

Now, I solved the test cases by sending SHUTDOWN NOW using a deferred client just before calling the `restart_server` proc. The `restart_server` and `kill_server` don't care if the server is already dead, so this works perfectly. I hope you agree. :)

Reply:
I don't think we need to run these test cases in both modes (both signal and shutdown now) because, again, all it affects is that assert.

Reply:
I'm paranoid that in some cases we send the GETACK and immediately exit (saving an RDB), imagine a case where the shutdown timeout is very short. 
In that case, that GETACK might be the thing that destroys the ability to PSYNC later. 

I know that we think currently there's no risk since e skip sending that GETACK when there's no need for it. But I still like to have these tests check both shutdown with and without timeout (I.e. NOW an and sigabrt)